Working Draft

alfiehanssen opened the pull request:
**Issue Description**  
Assemble a "one-page" overview of our product development process. Focus on central tenets / guidelines rather than tactics. The document should be easy for a newcomer to absorb "at a glance". It should be flexible enough so that there's room for others to interpret and apply as they see fit. The document should be publicly accessible (to all of Vimeo, and to the general public). 

**Implementation**  
I chose Github because it will allow us to evolve this document over time while maintaining a detailed history of its evolution, and because it will be accessible to all once we make this repo public.

I tried to strike a balance between text and graphics. Tried to be as concise as possible. Tried to balance the hard and soft aspects of our process. 

**Feedback**  
Please add comments inline or on the overall PR. Please do not edit the doc directly. I'll incorporate all feedback. Everything is fair game, including the repo title and diagrams (the sketch file is in the repo). If you'd like your name under the "contributors" section to link out to a webpage, please leave a comment with the URL you'd like to link to.

alfiehanssen committed:
Pasting in unformatted notes as a starting point.

alfiehanssen committed:
Started formatting, checking in part way.

alfiehanssen committed:
Pausing in order to commit images and Sketch file to repo.

alfiehanssen committed:
Added Sketch file and first diagram image to repo.

alfiehanssen committed:
Added team composition diagram.

alfiehanssen committed:
Adjusted team composition diagram subtitle.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Updated team composition image.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Some revisions and a footnote.

alfiehanssen committed:
Revised busted footnote.

alfiehanssen committed:
Made notes section header as large as others.

alfiehanssen committed:
Added a TOC.

alfiehanssen committed:
Revised TOC.

alfiehanssen committed:
Killed TOC

alfiehanssen committed:
Added list of contributors.

alfiehanssen committed:
Contributors list formatting was jacked.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Added sprint gant chart diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Added sprint diagram.

alfiehanssen committed:
Forgot to add image to vc.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Killed transparent bg.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Revised workflow diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Created sprint planning diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Added sprint planning diagram.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

kvenn commented:
###General 
- [x] I don't think I like the word co-location 
- [x] Checkmark instead of 100 for qa?
  - like them signing off as opposed to grading our work
- [x] Have someone outside the team read it for understandability before distribution
- [x] While the page itself is very high level and intentionally short, we could maybe link out to some other docs that further explain pieces
  - People that really want to know more can click into the link
  - An example of this would be linking to our "Velocity Tracking" page on confluence when the word "velocity" is used.

###Workflow Section
- [x] Provide an example maybe? it would go a long way

###Sprints Section
- [x] number the diagram below to match the numbers - it's hard to follow
- [x] describe what function takes care of each step in the written part
  - I think that's important
  - Actually that's hard because it could be QA, Design, or Dev (Anyone with tickets)
    - But it's a little unclear who's doing each step (is it a product manager? is it a project manager?)
- [x] Use a different word than "assign" for step 4
- [x] it was used in step 2

###Step 5
- [x] Commit to a specific subset of the ~~stories~~chunks? by using historical team velocity to calculate how much ~~the team~~ each person should take on
  - I've seen some people afraid of velocity because they think it's a per person metric
  - I think that sentence might be confusing to non-agiliers

###Retrospective
- [x] maybe mention that this is the time to customize the process to our needs (so people understand this is where processes can diverge and grow and they're not all the same)
- [x] Make ourselves more efficient - remove blockers

[EDIT]
- [x] In the area around velocity - mention how important predictability is. Everyone hates being blocked by a dependency and this process is best and reducing blockers.

nicolelehrer commented:
This is really cool Alfie. I second Kyle on the Sprint planning diagram - while the minimalism is really nice it's hard to gather at a glance the meaning behind the "clusters" without some labeling. Once you get it though, the visuals work well. I think labeling each column with a title would be great and think it's fair to be verbose when communicating something as complex as this in a graphic. e.g. "user stories", "assignment", "ticketing", "estimation", "commitment". I think the vertical axis could also use some clarification - here we're showing just 2 units of a 4 unit team composition. Can we clarify this with some labels as well? Or include all 4 groups? Or only show one team member as a representative per group if that's too much jazz. 

[EDIT]
- [x] Label the workflow columns e.g. "user stories", "assignment", "ticketing", "estimation", "commitment"
- [x] Workflow: I think the vertical axis could also use some clarification - here we're showing just 2 units of a 4 unit team composition. Can we clarify this with some labels as well? Or include all 4 groups? Or only show one team member as a representative per group if that's too much jazz.

howardrigberg commented:
- [x] It might help to begin with a background info section. Include a terse description about the basics of Scrum/Agile along with a (not overwhelming) link or two. We're using Scrum/Agile as a foundation, so it might be clearer to establish that first and then discuss our process in terms of how it adheres/departs/enhances typical Scrum/Agile.
- [x] The color coding of the workflow diagram is cool and makes the connection between who does what, but I'm not sure that everyone that reads the doc will pick up on that. Could still use some redundant labelling.
- [x] Notes bookmark is broken.
- [x] +1 on @kvenn's comment regarding the sprints diagram. It is cool and _almost_ there, but needs more clarity.
- [x] I agree with other commenters that a concrete example or two would go a long way. 
- [x] It seems like a following section or two could be added to get into more of the specifics of how we do things (e.g. mention daily stand up meetings, planning poker, etc). Or is that getting in the weeds too much?

fuzzywarblz commented:
would like to simplify the process diagram a bit...here's an idea...made some slight changes...gimme comments

![newnew 001](https://cloud.githubusercontent.com/assets/11947699/16130263/50980994-33d6-11e6-86cd-9d682ee2e5b4.png)

alfiehanssen commented:
@kvenn Couple questions: 
1. Why don't you like the term "co-location"? It means exactly what I'm trying to communicate. Do you have a suggestion for an alternative?
2. I was thinking of @dinasolovey's executing test cases and reporting the percent that passed/failed when I used the "100" emoji. Does that make it better? I was thinking about a checkmark and honestly I just thought the "100" looked better. 
3. We can certainly link out to specifics. Let me know if you have suggestions. It'd have to be someplace other than Confluence though so it's publicly accessible. For velocity tracking, we can link to Pace and communicate anything we want about velocity tracking in the Pace READMEs. I also like the idea of not connecting too much detail to this, so it doesn't become another 10 page handbook that's tough to maintain. I guess it depends on the info.

alfiehanssen committed:
Added a workflow example.

kvenn commented:
1) It was just a feeling in my heart. It felt like a mechanical word to describe a fluid concept. I wouldn't say it's necessary to change it. Maybe using softer language the second time like "being close by makes communication easier", or not.
2) 💯 feels like apple pie now. I'm 💯 percent on board.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

kvenn commented on README.md:
```diff
@@ -43,6 +43,8 @@ Teams evolve each idea one function at a time, with each function’s output bei
 <sup>*Workflow and deliverables*</sup>
 
+Consider an example: "As a user, I'd like to be able to upload videos to Vimeo from within the Vimeo iOS app." The team's product manager creates a document detailing the feature requirements. The team's designer uses this document to generate wireframes, visuals, etc. The team's developer uses the product and design documentation to build the upload feature. And the team's QA representative test ...
+
```
That sounds really good - maybe mentioning the sprint to sprint handoff somewhere in here so it's clear that after each sprint is a functional role. Or not since the diagram pretty much covers that

alfiehanssen committed:
Update README.md

alfiehanssen commented:
- [x] Around retrospective, discuss how burndown graph tells a story / can inform process tweaks.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Some image modifications.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Update README.md

alfiehanssen committed:
New diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Started working on sketch revisions.

alfiehanssen committed:
Revised sprint planning diagram.

alfiehanssen committed:
Diagram.

alfiehanssen committed:
Touch diagram.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Updates to sprint planning diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Updates to sprint planning diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Update README.md

alfiehanssen commented:
@kvenn I've asked Brian O'Hagan and Jake Oliver to review the doc. Will continue to solicit feedback from people outside the m&tv group.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen commented:
@fuzzywarblz Does the revised "sprint planning" diagram address some of your concerns?

alfiehanssen committed:
Update README.md

alfiehanssen commented:
@kvenn @fuzzywarblz @nicolelehrer @howardrigberg Can you give it another once over, tried to address all of your comments...

alfiehanssen committed:
Update README.md

alfiehanssen commented:
@toantronics Can you review and let me know what you think? I'd like to get a designer's perspective in the mix

alfiehanssen commented:
I'm shooting to merge this PR by Tuesday EOD, if you have time to add feedback by then... 👍 . If not, the review period is always open. And we can open separate PRs for anything and everything in the future.

nicolelehrer commented:
Looks awesome! I think it reads very easily, is at the right resolution, but provides more info for those who want it. Providing the specific upload example was a great idea. I think the diagram is also very clear while communicating so much, but still is playful and fun to look at. It's like an agile Noah's ark. 

Had some other thoughts, just food for thought maybe for a later update - Jon highlighted an interesting aspect in his diagram. The way I interpreted his 4 part square is that user stories need to be teased out of abstract ideas. Doing this requires (1) discussion across disciplines early, which increases chances of success/predictability down the chain (e.g. product wants X, design identifies potential conceptual inconsistencies, engineering identifies technical limitations, product updates - and so there's checks and balances that happen early so that we're less likely to encounter implementation specific problems down the chain.) But implementation specific problems are not always avoidable so there's also (2) a willingness to engage in iterative design / be adaptive across disciplines while we are mid-process. Design specs something out, engineering builds it and discovers something unanticipated, and design re-evaluates and updates and potentially uses this knowledge in future planning. (3) Communication across platforms also helps - solving across different constraints helps us abstract and define problems better.

howardrigberg commented:
Much improved! Really concise and to the point. I like it. Just a couple more suggestions. 
1. An important part of our process is figuring out how much we can accomplish as a team and being able to predict how long things will take in the short and long term. Assigning points to a story is one part of that (covered in doc), but determining the appropriate number of points to people for a sprint is the other important part (not covered in doc). In general, the theme of estimation and its value can be highlighted a bit more throughout the doc. As an example, I think you could fill out the description of what happens when we finish a sprint a little bit more. It would be helpful to talk a little bit about how we close out the sprint, carry out/ carry in, point estimation, etc. Some of that is in flux for us at the moment, but it is worth talking about regardless. 
2. Some of the single concepts in the notes section (e.g. scrum, agile, etc) could just be deep linked directly in the doc.

jasonhawkins commented on README.md:
```diff
@@ -1 +1,107 @@
+ 
+A product team is a co-located, cross-functional group of roughly 4-7 people who collaborate to bring ideas to life. Proximity builds relationships. Cross-functionality fosters autonomy and self-sufficiency. Limiting team size promotes ownership.
+ 
+![Team composition](https://github.com/vimeo/Blueprint/blob/working-draft/assets/team-composition.png)    
```
This image makes it look like Product and Dev don't interact, is that what you intended?

alfiehanssen replied:
Great point, see revised diagram

jasonhawkins commented on README.md:
```diff
@@ -1 +1,107 @@
+Workflow is a structured approach to how a team evolves an idea from inception to production. Teams evolve each idea one function at a time, with each function’s output being the next function’s input. 
+
+![Workflow and deliverables](https://github.com/vimeo/Blueprint/blob/working-draft/assets/workflow.png)    
+<sup>*Workflow and deliverables*</sup>
```
I think this is a good representation of our ideal workflow, but I think in practice it ends up being a lot of back and forth.

alfiehanssen replied:
Agreed, I think that's something that we can speak to if someone has questions about the doc, I'm also going to add some info addressing this per @nicolelehrer and @howardrigberg 's comment above

jasonhawkins commented on README.md:
```diff
@@ -1 +1,107 @@
+4. We estimate the amount of effort required to complete each chunk (via [planning poker](#notes)<sup>6</sup>), and
+5. We commit to a specific subset of chunks by using our historical team [velocity](#notes)<sup>7</sup> to calculate how much each person should take on.
+
+![Sprint planning](https://github.com/vimeo/Blueprint/blob/working-draft/assets/sprint-planning.png)    
```
hah, I like this a lot.

alfiehanssen committed:
Revised sprint planning diagram.

alfiehanssen committed:
Trying out SVG image.

alfiehanssen committed:
Update README.md

jasonhawkins commented:
This might be digging in too deep on the dev side, but I think peer review for our code is super important too. It might be worth calling out?

alfiehanssen committed:
PDF diagram.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Removed svg and pdf.

alfiehanssen committed:
Merge branch 'working-draft' of github.com:vimeo/Blueprint into working-draft

alfiehanssen committed:
Updated team composition diagram.

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Update README.md

alfiehanssen committed:
Updates sprint planning diagram, more spacing.

alfiehanssen committed:
Nothing

alfiehanssen committed:
Removed duplicate image.

alfiehanssen committed:
Update README.md

alfiehanssen commented:
@vimeo/mobile-and-tv-group 

I made a bunch of revisions hopefully addressing your feedback (excellent feedback!), 

I'm going to merge this branch, 

Retroactive feedback on the PR is definitely welcome, or feel free to make a new PR at any point,

ghking commented on README.md:
```diff
@@ -1 +1,107 @@
+
+Consider an example where the idea we're working on is captured by this user story: *"As a user, I'd like to be able to upload videos to Vimeo from within the Vimeo iOS app."* The team's product manager creates a feature spec. The team's designer uses this spec to generate visuals. The team's developer uses the product and design documentation to build the upload feature. And the team's QA repres ...
+
+Each function's output must be reviewed and approved by the other functions before it can advance to the next stage. This helps us make decisions when we have enough info to make them (not before, not after). It also ensures cross-functional ownership at each stage.
```
Maybe be more clear about what "other functions" means. Like all other functions? Or just adjacent ones?

alfiehanssen replied:
👍

ghking commented on README.md:
```diff
@@ -1 +1,107 @@
+
+This is our ideal workflow. In practice the boundaries are softer and the communication more fluid. No matter how the workflow is applied the emphasis is on discussion across functions and platforms early and often (e.g. product specs out a feature, design identifies potential conceptual inconsistencies, engineering identifies technical limitations, product updates the spec, and so on). 
+
+These checks and balances increase our chances of success and predictability by helping us abstract and define problems better and therefore making us less likely to encounter issues down the line. However, problems are not always avoidable and this is where a willingness to engage in iterative design is important.
```
Just design?

alfiehanssen replied:
👍

ghking commented on README.md:
```diff
@@ -1 +1,107 @@
+ 
+1. We assemble a rough list of stories to be included in the sprint (a mix of product, design, dev, and qa stories),
+2. We determine who will work on each story (via self-selection rather than top-down assignment), 
+3. Assignees break each of their stories into small, actionable chunks,
```
We could say "as tickets" unless thats too specific to Jira. Maybe add like "(as a ticket, for example)" to emphasize that this is a tangible chunk not just a mental chunk in the developers head

alfiehanssen replied:
👍

alfiehanssen commented:
@ghking I made edits based on your comments here: 

https://github.com/vimeo/Blueprint/commit/a17d68b820ea189a8df7e2865b909f882cb4d354
